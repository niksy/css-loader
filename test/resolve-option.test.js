import { join } from 'path';

import { webpack, evaluated, normalizeErrors } from './helpers';

describe('resolve option', () => {
  it('Function', async () => {
    const config = {
      loader: {
        options: {
          resolve: (resourcePath) => {
            expect(typeof resourcePath === 'string').toBe(true);
            if (resourcePath.includes('~package')) {
              return join('node_modules', resourcePath.replace('~', ''));
            }
            return resourcePath;
          },
        },
      },
    };
    const testId = './resolve/resolve.css';
    const stats = await webpack(testId, config);
    const { modules } = stats.toJson();
    const module = modules.find((m) => m.id === testId);

    expect(module.source).toMatchSnapshot('module');
    expect(evaluated(module.source, modules)).toMatchSnapshot(
      'module (evaluated)'
    );
    expect(normalizeErrors(stats.compilation.warnings)).toMatchSnapshot(
      'warnings'
    );
    expect(normalizeErrors(stats.compilation.errors)).toMatchSnapshot('errors');
  });

  it('CSS Modules', async () => {
    const config = {
      loader: {
        options: {
          modules: true,
          resolve: (resourcePath) => {
            expect(typeof resourcePath === 'string').toBe(true);
            if (resourcePath.includes('~@localpackage')) {
              return join('node_modules', resourcePath.replace('~', ''));
            }
            return resourcePath;
          },
        },
      },
    };
    const testId = './resolve/modules.css';
    const stats = await webpack(testId, config);
    const { modules } = stats.toJson();
    const module = modules.find((m) => m.id === testId);

    expect(module.source).toMatchSnapshot('module');
    expect(evaluated(module.source, modules)).toMatchSnapshot(
      'module (evaluated)'
    );
    expect(normalizeErrors(stats.compilation.warnings)).toMatchSnapshot(
      'warnings'
    );
    expect(normalizeErrors(stats.compilation.errors)).toMatchSnapshot('errors');
  });
});
