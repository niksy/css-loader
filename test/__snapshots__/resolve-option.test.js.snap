// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolve option CSS Modules: errors 1`] = `Array []`;

exports[`resolve option CSS Modules: module (evaluated) 1`] = `
Array [
  Array [
    2,
    "
",
    "",
  ],
  Array [
    1,
    "._2TC_ugPcaxHY-9vGyMb9K3 {
  color: gray;
}
",
    "",
  ],
]
`;

exports[`resolve option CSS Modules: module 1`] = `
"exports = module.exports = require(\\"../../../src/runtime/api.js\\")(false);
// Imports
exports.i(require(\\"-!../../../src/index.js??ref--4-0!./node_modules/@localpackage/color.css\\"), \\"\\");

// Module
exports.push([module.id, \\"._2TC_ugPcaxHY-9vGyMb9K3 {\\\\n  color: \\" + require(\\"-!../../../src/index.js??ref--4-0!./node_modules/@localpackage/color.css\\").locals[\\"color-grey\\"] + \\";\\\\n}\\\\n\\", \\"\\"]);

// Exports
exports.locals = {
	\\"color-grey\\": \\"\\" + require(\\"-!../../../src/index.js??ref--4-0!./node_modules/@localpackage/color.css\\").locals[\\"color-grey\\"] + \\"\\",
	\\"copyright\\": \\"_2TC_ugPcaxHY-9vGyMb9K3\\"
};"
`;

exports[`resolve option CSS Modules: warnings 1`] = `Array []`;

exports[`resolve option Function: errors 1`] = `Array []`;

exports[`resolve option Function: module (evaluated) 1`] = `
Array [
  Array [
    2,
    ".test {
  a: a;
}
",
    "",
  ],
  Array [
    3,
    ".test {
  a: a;
}
",
    "screen and print",
  ],
  Array [
    5,
    "a {
  b: b;
}
",
    "((min-width: 100px)) and (screen and print)",
  ],
  Array [
    4,
    ".test {
  c: c;
}
",
    "screen and print",
  ],
  Array [
    1,
    "@import url(http://example.com/style.css);",
    "",
  ],
  Array [
    1,
    "@import url(http://example.com/style.css#hash);",
    "",
  ],
  Array [
    1,
    "@import url(http://example.com/style.css?#hash);",
    "",
  ],
  Array [
    1,
    "@import url(http://example.com/style.css?foo=bar#hash);",
    "",
  ],
  Array [
    1,
    "@import url(http://example.com/other-style.css);",
    "screen and print",
  ],
  Array [
    1,
    "@import url(//example.com/style.css);",
    "",
  ],
  Array [
    6,
    ".test {
  d: d
}
",
    "",
  ],
  Array [
    7,
    ".query {
  e: e;
}
",
    "",
  ],
  Array [
    8,
    ".other-query {
  f: f;
}
",
    "",
  ],
  Array [
    9,
    ".other-query {
  f: f;
}
",
    "screen and print",
  ],
  Array [
    1,
    "@import url(https://fonts.googleapis.com/css?family=Roboto);",
    "",
  ],
  Array [
    1,
    "@import url(https://fonts.googleapis.com/css?family=Noto+Sans+TC);",
    "",
  ],
  Array [
    1,
    "@import url(https://fonts.googleapis.com/css?family=Noto+Sans+TC|Roboto);",
    "",
  ],
  Array [
    10,
    ".relative {
  color: red;
}
",
    "",
  ],
  Array [
    11,
    ".top-relative {
  color: black;
}
",
    "",
  ],
  Array [
    12,
    ".tilde {
  color: yellow;
}
",
    "",
  ],
  Array [
    13,
    ".alias {
  color: red;
}
",
    "",
  ],
  Array [
    14,
    ".background-imported {
  background: url(/webpack/public/path/img.png);
}
",
    "",
  ],
  Array [
    1,
    "@import url();
@import url('');
@import url(\\"\\");
@import '';
@import \\"\\";
@import \\"   \\";
@import \\"
\\";
@import url();
@import url('');
@import url(\\"\\");
@import ;
@import foo-bar;
@import-normalize;
@import url('http://') :root {}

.class {
  a: b c d;
}

.foo {
  @import 'path.css';
}

.background {
  background: url(/webpack/public/path/img.png);
}
",
    "",
  ],
]
`;

exports[`resolve option Function: module 1`] = `
"exports = module.exports = require(\\"../../../src/runtime/api.js\\")(false);
// Imports
exports.i(require(\\"-!../../../src/index.js??ref--4-0!./test.css\\"), \\"\\");
exports.i(require(\\"-!../../../src/index.js??ref--4-0!./test.css\\"), \\"screen and print\\");
exports.i(require(\\"-!../../../src/index.js??ref--4-0!./test-media.css\\"), \\"screen and print\\");
exports.i(require(\\"-!../../../src/index.js??ref--4-0!./test-other.css\\"), \\"(min-width: 100px)\\");
exports.push([module.id, \\"@import url(http://example.com/style.css);\\", \\"\\"]);
exports.push([module.id, \\"@import url(http://example.com/style.css#hash);\\", \\"\\"]);
exports.push([module.id, \\"@import url(http://example.com/style.css?#hash);\\", \\"\\"]);
exports.push([module.id, \\"@import url(http://example.com/style.css?foo=bar#hash);\\", \\"\\"]);
exports.push([module.id, \\"@import url(http://example.com/other-style.css);\\", \\"screen and print\\"]);
exports.push([module.id, \\"@import url(//example.com/style.css);\\", \\"\\"]);
exports.i(require(\\"-!../../../src/index.js??ref--4-0!./node_modules/package/test.css\\"), \\"\\");
exports.i(require(\\"-!../../../src/index.js??ref--4-0!./query.css?foo=1&bar=1\\"), \\"\\");
exports.i(require(\\"-!../../../src/index.js??ref--4-0!./other-query.css?foo=1&bar=1#hash\\"), \\"\\");
exports.i(require(\\"-!../../../src/index.js??ref--4-0!./other-query.css?foo=1&bar=1#hash\\"), \\"screen and print\\");
exports.push([module.id, \\"@import url(https://fonts.googleapis.com/css?family=Roboto);\\", \\"\\"]);
exports.push([module.id, \\"@import url(https://fonts.googleapis.com/css?family=Noto+Sans+TC);\\", \\"\\"]);
exports.push([module.id, \\"@import url(https://fonts.googleapis.com/css?family=Noto+Sans+TC|Roboto);\\", \\"\\"]);
exports.i(require(\\"-!../../../src/index.js??ref--4-0!./relative.css\\"), \\"\\");
exports.i(require(\\"-!../../../src/index.js??ref--4-0!../import/top-relative.css\\"), \\"\\");
exports.i(require(\\"-!../../../src/index.js??ref--4-0!./node_modules/package/tilde.css\\"), \\"\\");
exports.i(require(\\"-!../../../src/index.js??ref--4-0!aliasesImport/alias.css\\"), \\"\\");
exports.i(require(\\"-!../../../src/index.js??ref--4-0!./url.css\\"), \\"\\");
var urlEscape = require(\\"../../../src/runtime/url-escape.js\\");
var ___CSS_LOADER_URL___0___ = urlEscape(require(\\"./img.png\\"));

// Module
exports.push([module.id, \\"@import url();\\\\n@import url('');\\\\n@import url(\\\\\\"\\\\\\");\\\\n@import '';\\\\n@import \\\\\\"\\\\\\";\\\\n@import \\\\\\"   \\\\\\";\\\\n@import \\\\\\"\\\\n\\\\\\";\\\\n@import url();\\\\n@import url('');\\\\n@import url(\\\\\\"\\\\\\");\\\\n@import ;\\\\n@import foo-bar;\\\\n@import-normalize;\\\\n@import url('http://') :root {}\\\\n\\\\n.class {\\\\n  a: b c d;\\\\n}\\\\n\\\\n.foo {\\\\n  @import 'path.css';\\\\n}\\\\n\\\\n.background {\\\\n  background: url(\\" + ___CSS_LOADER_URL___0___ + \\");\\\\n}\\\\n\\", \\"\\"]);

"
`;

exports[`resolve option Function: warnings 1`] = `
Array [
  "ModuleWarning: Module Warning (from \`replaced original path\`):
Warning

(12:1) Unable to find uri in '@import url()'",
  "ModuleWarning: Module Warning (from \`replaced original path\`):
Warning

(13:1) Unable to find uri in '@import url('')'",
  "ModuleWarning: Module Warning (from \`replaced original path\`):
Warning

(14:1) Unable to find uri in '@import url(\\"\\")'",
  "ModuleWarning: Module Warning (from \`replaced original path\`):
Warning

(17:1) Unable to find uri in '@import '''",
  "ModuleWarning: Module Warning (from \`replaced original path\`):
Warning

(18:1) Unable to find uri in '@import \\"\\"'",
  "ModuleWarning: Module Warning (from \`replaced original path\`):
Warning

(19:1) Unable to find uri in '@import \\"   \\"'",
  "ModuleWarning: Module Warning (from \`replaced original path\`):
Warning

(20:1) Unable to find uri in '@import \\"
\\"'",
  "ModuleWarning: Module Warning (from \`replaced original path\`):
Warning

(22:1) Unable to find uri in '@import url()'",
  "ModuleWarning: Module Warning (from \`replaced original path\`):
Warning

(23:1) Unable to find uri in '@import url('')'",
  "ModuleWarning: Module Warning (from \`replaced original path\`):
Warning

(24:1) Unable to find uri in '@import url(\\"\\")'",
  "ModuleWarning: Module Warning (from \`replaced original path\`):
Warning

(40:1) Unable to find uri in '@import '",
  "ModuleWarning: Module Warning (from \`replaced original path\`):
Warning

(41:1) Unable to find uri in '@import foo-bar'",
  "ModuleWarning: Module Warning (from \`replaced original path\`):
Warning

(43:1) It looks like you didn't end your @import statement correctly. Child nodes are attached to it.",
]
`;
